lib/ansible/playbook/conditional.py

```diff
@@ -213,14 +213,28 @@ def generic_visit(self, node, inside_call=False, inside_yield=False):
             # to be looking for an undefined variable, return True, otherwise fail
             try:
                 # first we extract the variable name from the error message
-                var_name = re.compile(r"'(hostvars\[.+\]|[\w_]+)' is undefined").search(str(e)).groups()[0]
+                found = re.compile(r"'(hostvars\[.+\]|[\w_]+)' is undefined").search(str(e))
+                if not found:
+                    var_name = None
+                else:
+                    var_name = found.groups()[0]
+
                 # next we extract all defined/undefined tests from the conditional string
                 def_undef = self.extract_defined_undefined(conditional)
+                # undefined var_name found, but conditional had no matches; likely a complex var
+                if not def_undef and var_name and ('is defined' in conditional or 'is not undefined' in conditional):
```

Comment:
wont this also match `when: not varname is defined` and not match `when: not varname is undefined`?